trieste/models/gpflow/models.py

```diff
@@ -1362,3 +1368,259 @@ def covariance_between_points(
+    All the base models used in this implementation are :class:`~gpflow.models.GPR` models.
+    Note: Currently only supports single output problems.
+    """
+
+    def __init__(
+        self,
+        fidelity_models: Sequence[GaussianProcessRegression],
+    ):
+        """
+        :param fidelity_models: List of
+            :class:`~trieste.models.gpflow.models.GaussianProcessRegression`
+            models, one for each fidelity.
+        """
+        self.num_fidelities = len(fidelity_models)
+
+        self.lowest_fidelity_signal_model = fidelity_models[0]
+        # Note: The 0th index in the below is not a residual model, and should not be used.
+        self.fidelity_residual_models: Sequence[GaussianProcessRegression] = fidelity_models
+        # set this as a Parameter so that we can optimize it
+        rho = [
+            gpflow.Parameter(1.0, trainable=True, name=f"rho_{i}")
+            for i in range(self.num_fidelities - 1)
+        ]
+        self.rho: list[gpflow.Parameter] = [
+            gpflow.Parameter(1.0, trainable=False, name="dummy_variable"),
+            *rho,
+        ]
+
+    def predict(self, query_points: TensorType) -> tuple[TensorType, TensorType]:
+        """
+        Predict the marginal mean and variance at query_points.
+
+        :param query_points: Query points with shape [N, D+1], where the
+            final column of the final dimension contains the fidelity of the query point
+        :return: mean: The mean at query_points with shape [N, P],
+                 and var: Th variance at query_points with shape [N, P]
+        """
+        (
+            query_points_wo_fidelity,
+            query_points_fidelity_col,
+        ) = check_and_extract_fidelity_query_points(
+            query_points, max_fidelity=self.num_fidelities - 1
+        )
+
+        signal_mean, signal_var = self.lowest_fidelity_signal_model.predict(
+            query_points_wo_fidelity
+        )
+
+        for fidelity in range(self.num_fidelities):
+            if fidelity == 0:
+                fidelity_residual_mean = 0
+                fidelity_residual_var = 0
+            else:
+                fidelity_residual_mean, fidelity_residual_var = self.fidelity_residual_models[
+                    fidelity
+                ].predict(query_points_wo_fidelity)
+
+            new_fidelity_signal_mean = self.rho[fidelity] * signal_mean + fidelity_residual_mean
+            new_fidelity_signal_var = fidelity_residual_var + (self.rho[fidelity] ** 2) * signal_var
+
+            mask = query_points_fidelity_col >= fidelity
+            signal_mean = tf.where(mask, new_fidelity_signal_mean, signal_mean)
+            signal_var = tf.where(mask, new_fidelity_signal_var, signal_var)
+
+        return signal_mean, signal_var
+
+    def _calculate_residual(self, dataset: Dataset, fidelity: int) -> TensorType:
+        """
+        Calculate the true residuals for a set of datapoints at a given fidelity.
```

Comment:
"true residuals" to me means purely data based. Might be good to add the equation here of what's computed for clarity?